### PR TITLE
[Dist] Add ZeRO-1 Optimizer

### PR DESCRIPTION
**Pull request by hgt312**
## Intro

A simple ZeRO-1 implementation in for torch-xla. Similar API with pytorch's: https://github.com/pytorch/pytorch/blob/master/torch/distributed/optim/zero_redundancy_optimizer.py, but different implmentation.

PyTorch use the legacy logic in [DeepSpeed](https://github.com/microsoft/DeepSpeed/blob/v0.5.0/deepspeed/runtime/zero/stage1.py), this approach separates params in several groups, and each rank is responsible for a group. This will make the workload on each rank be different and may be unbalanced.
We are proposing to use the updated logic in [DeepSpeed](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/stage_1_and_2.py), which split/slice each tensor to world size and each rank is responsible for a partition, so that the workload on each are same. Furthermore, we use `reduce_scatter` instead of `all_reduce + split/chunk/slice` and shard params on CPU firstly to reduce generated graphs and achieve SPMD.

## Usage

Just wrap the optimizer:

```python
from torch_xla.distributed.zero_redundancy_optimizer import ZeroRedundancyOptimizer

optimizer_wrapper = ZeROStage1(model.parameters(), SGD, ...)
xm.mark_step()
for ...:
    loss = model(inputs)
    loss.backward()
    xm.mark_step()
    ...
    optimizer_wrapper.step()
    xm.mark_step()
    ...
```

### About ZeRO-1, see https://arxiv.org/abs/1910.02054

```
ZeRO-1 workflow (underlined parts are core logics): 
model -> all-reduce grads of parameters ->
shard grads and assign to pre-sharded parameters -> optimizer ->
all-gather sharded parameters
```

### Status
- test on colab, it works
- with an option to enable/disable grad clipping
- with an option to determine optimizer dtype
- missing functionality: padding

### CODE REVIEW

**Comment by jspisak** [comment]
@mrshenli - any thoughts here on usage and whether to land this in torchxla?

File: torch_xla/distributed/zero_redundancy_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        )
+
+    return loss
+
+  def state_dict(self):
+    state_dict = super().state_dict()
+    state_dict['base'] = self.base_optimizer.state_dict()
+    return state_dict
+
+  def load_state_dict(self, state_dict):
```
Comment (mrshenli) [inline torch_xla/distributed/zero_redundancy_optimizer.py:153]: do we need test coverage for `state_dict` and `load_state_dict`?

**Review by mrshenli — COMMENTED** [review]
Sorry about the delay. LGTM! Left some minor comments. Question, will FSDP ZeRO2 always performs better than doing this in the optimizer? The former also runs reduce_scatter + all_gather, but those comm ops can better overlap with forward and backward computations. 

Hey @wconstab, is there any items that we need to check before landing?

File: torch_xla/distributed/zero_redundancy_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    tensor = tensor.chunk(self.world_size)[self.rank]
+    return tensor
+
+  def _shard_parameters(self):
+    """
+        Shard all parameters.
+        """
+    xm.unlazy(self.params)
+    for param in self.params:
+      shard_data = param.data.to(device="cpu")  # move to cpu
```
[the remark this answers is not in the document]
Comment (hgt312) [reply torch_xla/distributed/zero_redundancy_optimizer.py:86]: It's a mistake that it has no impact on num of graphs (1 vs 1).
In our use case, usually there is a `mark_step` after model init, this will generate one graph:
```python
model, optimizer = init_model_and_optimizer(...)
xm.mark_step()
```
as shard depends on rank, if we shard on xla device, each rank will generate a different graph. by moving this to cpu, we can share only one compiled graph between processes

File: torch_xla/distributed/zero_redundancy_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+          grad_shard = grad_shard.to(dtype=self.optimizer_dtype)
+        shard.grad = grad_shard
+
+    if self.grad_clipping:
+      # Update unscale/clip with sub partitions
+      torch.nn.utils.clip_grad_norm_(
+          self.sharded_params, max_norm=self.max_norm)
+
+    # Step the wrapped optimizer
+    loss = self.base_optimizer.step(closure=closure, **kwargs)
```
[the remark this answers is not in the document]
Comment (hgt312) [reply torch_xla/distributed/zero_redundancy_optimizer.py:128]: thanks! get it, will add soon

File: torch_xla/distributed/zero_redundancy_optimizer.py (hunk 1 above)
Comment (hgt312) [reply torch_xla/distributed/zero_redundancy_optimizer.py:153]: yes, i'll add them

**Comment by hgt312** [comment]
> Sorry about the delay. LGTM! Left some minor comments. Question, will FSDP ZeRO2 always performs better than doing this in the optimizer? The former also runs reduce_scatter + all_gather, but those comm ops can better overlap with forward and backward computations.
> 
> Hey @wconstab, is there any items that we need to check before landing?

In our use cases, the scripts are like:
```python
for ...:
    loss = model(inputs)
    loss.backward()
    xm.mark_step()
    if ...:  # meet grad acc boundary
        optimizer_wrapper.step()
        xm.mark_step()
    ...
```
fwd+bwd in one graph, optimizer and some misc (norm, grad clipping) in one graph, and the two graphs are not overlapped

for the performance comparison between ZeRO-1 and ZeRO-2, for some model config, ZeRO-1 is better on GPU
for xla, we have not test them and do not have a ZeRO-2 implementation for xla now

we'd like to use FSDP ZeRO-2 but FSDP in xla only have ZeRO-3

**Comment by wconstab** [comment]
i'm wondering why this lands in pytorch/pytorch instead of pytorch/xla - after reading the code it seems specific to xla.

otherwise, it looks ok to me.  (Similar questions as @hgt312 about perf, i'm curious what the overall plan for this is and whether someone is also working on a version with overlap).

**Comment by JackCaoG** [comment]
@wconstab this pr is going to land to pytorch/xla, through the branch name [pytorch:master](https://github.com/pytorch/xla) is a bit misleading.

**Comment by JackCaoG** [comment]
I will land this change to the master, but I will most likely leave this one out of the 2.0 release.

**Comment by hgt312** [comment]
> I will land this change to the master, but I will most likely leave this one out of the 2.0 release.

Thanks for your reply! Which version do you consider this?

And do your have the plan to update FSDP in ptxla to align FSDP in pytorch to support more features?

**Comment by JackCaoG** [comment]
It will most likely be in the 2.1 release, we are shifting our development focus to the SPMD/DTensor. We will maintain the FSDP code and add features as user request. This might change when the compiler version of the FSDP is being developed on the upstream, hopefully we can share the same implementation by then.
